nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/smtp/handler/SMTPMessageHandlerFactory.java

```diff
@@ -0,0 +1,147 @@
+            this.from = from;
+        }
+
+        @Override
+        public void recipient(String recipient) throws RejectException {
+            // TODO: possibly whitelist receivers?
+            this.recipient = recipient;
+        }
+
+        @Override
+        public void data(InputStream inputStream) throws RejectException, TooMuchDataException, IOException {
+            // Start counting the timer...
+
+            StopWatch watch = new StopWatch(false);
+
+            SMTPServer server = messageContext.getSMTPServer();
+
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+            byte [] buffer = new byte[1024];
+            int rd;
+
+            while ((rd = inputStream.read(buffer, 0, buffer.length)) != -1) {
+                baos.write(buffer, 0, rd);
+            }
+            if (baos.getBufferLength() > server.getMaxMessageSize()) {
+                throw new TooMuchDataException("Data exceeds the amount allowed.");
+            }
+
+            baos.flush();
+            this.messageBody = baos.toByteArray();
+
+
+            X509Certificate[] certificates = new X509Certificate[]{};
+
+            String remoteIP = messageContext.getRemoteAddress().toString();
+            String helo = messageContext.getHelo();
+            if (messageContext.getTlsPeerCertificates() != null ){
+                certificates = (X509Certificate[]) messageContext.getTlsPeerCertificates().clone();
+            }
+
+            SmtpEvent message = new SmtpEvent(remoteIP, helo, from, recipient, certificates, messageBody);
+            try {
+                    // Try to queue the message back to the NiFi session
+                    incomingMessages.put(message);
+            } catch (InterruptedException e) {
+                // Throws an error to the  incoming server alerting about issues queuing the message in NiFi
+                logger.error("Hit an error sending message back to NiFi main thread {}. Sending and error back to SMTP client and discarding message", new Object [] {e});
+                throw new DropConnectionException(451, "NiFi something went wrong while processing you message, please retry again later");
+            }
+
+            // Once message has been sent to the queue, it should be processed by NiFi onTrigger,
+            // a flowfile created and its processed status updated before an acknowledgment is
+            // given back to the SMTP client
+            try {
+                synchronized(message) {
+                    while(! message.getProcessed()) {
+                        // Check to see if it is too late...
+                        final long serverTimeout = TimeUnit.NANOSECONDS.convert(messageContext.getSMTPServer().getConnectionTimeout(), TimeUnit.MILLISECONDS);
+                        if ( watch.getElapsed(TimeUnit.MILLISECONDS) <=  serverTimeout) {
+                            // Will should allow for a couple of retries;
```

Comment:
How is this retrying? From what I can tell, after the message is put/offered to the queue, the OnTrigger will take it off and if it fails it just hangs in limbo (until the server timeout happens). 

You could probably go one of two routes:
1: No retries, after the message is taken off the queue the processor attempts to process it. If it fails, the server doesn't try to reprocess it and replies with a 451. To do this, you should add a flag to SmtpEvent with a failure flag/string to convey that the message failed to process. That way you can fail fast (instead of timing out).

2: Retries, the processor "peeks" at the queue instead of taking messages off of it. Then once it has processed the message it removes it from the queue. This would require a lock over the queue and for the processor to trigger serially and probably changes in other places as well. This is how ConsumeMQTT works. 

I like option 1, it doesn't require significantly changing your code and easily fits with the SMTP protocol.